src/libraries/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -125,27 +102,47 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, Index
             return element;
         }
 
-        private static bool TryGetElement<TSource>(IEnumerable<TSource> source, int index, [MaybeNullWhen(false)] out TSource element)
+        private static TSource? TryGetElementAt<TSource>(this IEnumerable<TSource> source, int index, out bool found, bool guardIListLength = true) =>
+#if !OPTIMIZE_FOR_SIZE
+            source is Iterator<TSource> iterator ? iterator.TryGetElementAt(index, out found) :
```

Comment:
One common micro-optimization we have been applying in STJ is avoiding type checks against generic types. Could a similar approach give dividends here? (e.g. checking against a non-generic `Iterator` base type and then directly casting it to the generic variant?)

Reply:
Potentially? We could try it and measure it after this PR.

Reply:
I gave it a shot but it seems IEnumerable being covariant makes such an optimization impossible in this context.